compiler/rustc_borrowck/src/polonius/typeck_constraints.rs

```diff
@@ -35,7 +41,191 @@ pub(super) fn convert_typeck_constraints<'tcx>(
+    match &stmt.kind {
+        StatementKind::Assign(box (lhs, rhs)) => {
+            // To create localized outlives constraints without midpoints, we rely on the property
+            // that no input regions from the RHS of the assignment will flow into themselves: they
+            // should not appear in the output regions in the LHS. We believe this to be true by
+            // construction of the MIR, via temporaries, and assert it here.
+            //
+            // We think we don't need midpoints because:
+            // - every LHS Place has a unique set of regions that don't appear elsewhere
+            // - this implies that for them to be part of the RHS, the same Place must be read and
+            //   written
+            // - and that should be impossible in MIR
+            //
+            // When we have a more complete implementation in the future, tested with crater, etc,
+            // we can relax this to a debug assert instead, or remove it.
+            assert!(
+                {
+                    let mut lhs_regions = FxHashSet::default();
+                    tcx.for_each_free_region(lhs, |region| {
+                        let region = universal_regions.to_region_vid(region);
+                        lhs_regions.insert(region);
+                    });
+
+                    let mut rhs_regions = FxHashSet::default();
+                    tcx.for_each_free_region(rhs, |region| {
+                        let region = universal_regions.to_region_vid(region);
+                        rhs_regions.insert(region);
+                    });
+
+                    // The intersection between LHS and RHS regions should be empty.
+                    lhs_regions.is_disjoint(&rhs_regions)
+                },
+                "there should be no common regions between the LHS and RHS of an assignment"
+            );
+
+            // As mentioned earlier, we should be tracking these better upstream but: we want to
+            // relate the types on entry to the type of the place on exit. That is, outlives
+            // constraints on the RHS are on entry, and outlives constraints to/from the LHS are on
+            // exit (i.e. on entry to the successor location).
+            let lhs_ty = body.local_decls[lhs.local].ty;
+            let successor_location = Location {
+                block: current_location.block,
+                statement_index: current_location.statement_index + 1,
+            };
+            let successor_point = liveness.point_from_location(successor_location);
+            compute_constraint_direction(
+                tcx,
+                outlives_constraint,
+                &lhs_ty,
+                current_point,
+                successor_point,
+                universal_regions,
+            )
+        }
+        _ => {
+            // For the other cases, we localize an outlives constraint to where it arises.
+            (current_point, current_point)
```

Comment:
To add to this, are these actually useful? (I think I remember that these get filtered somewhere? Or was that something else?)

If these *aren't* useful, then you can make this branch empty.